lib/service/awsoidc.go

```diff
@@ -0,0 +1,414 @@
+			updater.Log.WithError(err).Debug("Deploy service agent update is already being processed.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := updater.AuthClient.CancelSemaphoreLease(ctx, *semLock); err != nil {
+			updater.Log.WithError(err).Error("Failed to cancel semaphore lease.")
+		}
+	}()
+
+	updater.Log.Debugf("Updating Deploy Service Agents for integration %s in AWS region: %s", integration.GetName(), awsRegion)
+	if err := awsoidc.UpdateDeployServiceAgent(ctx, deployServiceClient, awsoidc.UpdateServiceRequest{
+		TeleportClusterName: updater.TeleportClusterName,
+		TeleportVersionTag:  teleportVersion,
+		OwnershipTags:       ownershipTags,
+	}); err != nil {
+
+		switch {
+		case trace.IsNotFound(err):
+			// The updater checks each integration/region combination, so
+			// there will be regions where there is no ECS cluster deployed
+			// for the integration.
+			updater.Log.WithError(err).Debugf("Integration %s does not manage any services within region %s.", integration.GetName(), awsRegion)
+			return nil
+		case trace.IsAccessDenied(awslib.ConvertIAMv2Error(trace.Unwrap(err))):
+			// The aws oidc role may lack permissions due to changes in teleport.
+			// In this situation users should be notified that they will need to
+			// re-run the deploy service iam configuration script and update the
+			// permissions.
+			updater.Log.WithError(err).Warning("Re-run deploy service configuration script to update permissions.")
+			if err := updater.ensureClusterAlert(ctx, integration, awsRegion); err != nil {
+				updater.Log.WithError(err).Warning("Failed to ensure cluster alert.")
+			}
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ensureClusterAlert ensures a cluster alert is created if deploy service permissions
+// need to be reconfigured.
+func (updater *DeployServiceUpdater) ensureClusterAlert(ctx context.Context, integration types.Integration, awsRegion string) error {
+	// Acquire semaphore lease before attempting to create a cluster alert.
+	// If the lease cannot be acquired, a cluster alert is already being created by another instance.
+	semLock, err := updater.AuthClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+		SemaphoreKind: types.SemaphoreKindConnection,
+		SemaphoreName: "aws_oidc_access_denied_alert",
+		MaxLeases:     1,
+		Expires:       updater.Clock.Now().Add(updateDeployAgentsInterval),
+		Holder:        "aws_oidc_access_denied_alert",
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), teleport.MaxLeases) {
+			updater.Log.WithError(err).Debug("Cluster alert is already being created.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := updater.AuthClient.CancelSemaphoreLease(ctx, *semLock); err != nil {
+			updater.Log.WithError(err).Error("Failed to cancel semaphore lease.")
+		}
+	}()
+
+	alertLabels := map[string]string{
+		types.AlertAWSOIDCAccessDenied: "yes",
+		types.AlertOnLogin:             "yes",
```

Comment:
I don't know if we need to alert on login. Every user will see it, and most of them won't be able to do anything about it.